### ISSUE TITLE
vststasks hangs when using impact analysis

### ISSUE THREAD

**Issue by henning-krause** [Microsoft/vsts-tasks#6842]
## Environment
- Server - TFS 2018.2rc on-premises
- Agent - Private, 2.131.0 on Windows Server 2016

## Issue Description

Task         : Visual Studio Test
Version      : 2.3.27

The tests in our solution are all executed successful. However, the last step (uploading test results) simply hangs and is eventually canceled when the build job times out.

When I disable the "Rerun failed tests" option, the task does not hang, but fails with this error:

```
Exception : Violation of PRIMARY KEY constraint 'PK__#B09096F__DD1FEA8DE20C9B6D'. Cannot insert duplicate key in object 'dbo.@resultsTable'. The duplicate key value is (103776).
The data for table-valued parameter "@resultsTable" doesn't conform to the table type of the parameter. SQL Server error is: 3602, state: 30
The statement has been terminated.
```

Note that I already had this last issue (SQL error) with the previous TFS version (2018 Update 1).

### Error logs

The following error logs where collected for the hanging issue:

Task parameters
```
2018-03-30T07:02:42.0006649Z ==============================================================================
2018-03-30T07:02:45.3447085Z Run the tests locally using vstest.console.exe
2018-03-30T07:02:45.3447477Z ========================================================
2018-03-30T07:02:45.3451423Z Test selector : Test assemblies
2018-03-30T07:02:45.3451894Z Test assemblies : **\NetatWork\**\*tests*.dll,**\Solutions\**\*tests*.dll,!**\obj\**
2018-03-30T07:02:45.3452308Z Test filter criteria : null
2018-03-30T07:02:45.3452679Z Search folder : C:\BuildAgent1\_work\45\s
2018-03-30T07:02:45.3453076Z Run settings file : C:\BuildAgent1\_work\45\s
2018-03-30T07:02:45.3453550Z Run in parallel : false
2018-03-30T07:02:45.3453915Z Run in isolation : false
2018-03-30T07:02:45.3456072Z Path to custom adapters : null
2018-03-30T07:02:45.3456444Z Other console options : null
2018-03-30T07:02:45.3456795Z Code coverage enabled : true
2018-03-30T07:02:45.3457495Z Rerun failed tests: true
2018-03-30T07:02:45.3458020Z Rerun failed tests threshold: 30
2018-03-30T07:02:45.3458370Z Rerun maximum attempts: 3
2018-03-30T07:02:45.3458870Z VisualStudio version selected for test execution : latest
```

```
2018-03-30T07:28:51.8889235Z Attachments:
2018-03-30T07:28:51.8900260Z   C:\BuildAgent1\_work\45\s\TestResults\05a7deb9-516b-4d8f-9ed5-2beaece98744\tfssrv_srv 2018-03-30 09_06_57.coverage
2018-03-30T07:28:51.8902034Z Total tests: 4245. Passed: 4222. Failed: 0. Skipped: 23.
2018-03-30T07:28:51.8902562Z Test Run Successful.
2018-03-30T07:28:51.9039151Z Test execution time: 21,8973 Minutes
2018-03-30T07:28:53.5832020Z Results File: C:\BuildAgent1\_work\45\s\TestResults\tfssrv_srv_2018-03-30_09_07_18.trx
2018-03-30T07:28:53.6180106Z **************** Completed test execution *********************
2018-03-30T07:28:53.6318001Z [command]C:\BuildAgent1\_work\_tasks\VSTest_ef087383-ee5e-42c7-9a53-ab56c98420f9\2.3.27\TestSelector\TestSelector.exe UpdateTestResults
2018-03-30T08:32:36.4405773Z ##[error]The operation was canceled.
2018-03-30T08:32:36.4467538Z ##[section]Finishing: Run Unittests
```

This I found in the worker logs:
```
[2018-03-30 07:28:50Z INFO JobServerQueue] Try to append 1 batches web console lines, success rate: 1/1.
[2018-03-30 07:28:51Z INFO JobServerQueue] Try to append 1 batches web console lines, success rate: 1/1.
[2018-03-30 07:28:52Z INFO JobServerQueue] Try to append 1 batches web console lines, success rate: 1/1.
[2018-03-30 07:28:53Z INFO JobServerQueue] Try to append 1 batches web console lines, success rate: 1/1.
[2018-03-30 08:32:35Z INFO Worker] Cancellation/Shutdown message received.
[2018-03-30 08:32:35Z INFO ExpressionManager] Evaluating: succeeded()
[2018-03-30 08:32:35Z INFO ExpressionManager] Expanded: False
[2018-03-30 08:32:35Z INFO ExpressionManager] Result: False
[2018-03-30 08:32:35Z INFO StepsRunner] Cancel current running step.
[2018-03-30 08:32:36Z INFO ProcessInvoker] Sending CTRL_C to process 8592.
[2018-03-30 08:32:36Z INFO ProcessInvoker] Successfully send CTRL_C to process 8592.
[2018-03-30 08:32:36Z INFO ProcessInvoker] Waiting for process exit or 10 seconds after CTRL_C signal fired.
[2018-03-30 08:32:36Z INFO ProcessInvoker] Ignore Ctrl+C to current process.
[2018-03-30 08:32:36Z INFO ProcessInvoker] Finished process with exit code -1073741510, and elapsed time 01:29:53.9729293.
[2018-03-30 08:32:36Z ERR  StepsRunner] Caught cancellation exception from step: System.OperationCanceledException: The operation was canceled.
```

**Comment by ShreyasRmsft** [comment]
@henning-krause can you please upload the trx files here? (**For both hang issue as well as non rerun scenario**)

(this one: Results File: C:\BuildAgent1\_work\45\s\TestResults\tfssrv_srv_2018-03-30_09_07_18.trx)


Also can you set system.debug to true and provide the full zip file containing all the logs for both scenarios?

**Comment by ShreyasRmsft** [comment]
@henning-krause any updates here?

**Comment by henning-krause** [comment]
I wasn't in the office during the last couple of day.. will upload the stuff ASAP.

**Comment by henning-krause** [comment]
@ShreyasRmsft, can I send you the logs via email?

**Comment by ShreyasRmsft** [comment]
@henning-krause please send the logs to devops_tools@microsoft.com

**Comment by ShreyasRmsft** [comment]
and the trx file too

**Comment by henning-krause** [comment]
@ShreyasRmsft: Done

**Comment by henning-krause** [comment]
@ShreyasRmsft, hmm.. I got  this error when sending the mail:

```
Gemeldeter Fehler:	 550 5.7.133 RESOLVER.RST.SenderNotAuthenticatedForGroup; authentication required; Delivery restriction check failed because the sender was not authenticated when sending to this group 	
DSN generiert von:	 CY1PR21MB0070.namprd21
```

But the error wasn't generated for devops_tools but for a different address. Is there a forwarding issue?

**Comment by ShreyasRmsft** [comment]
@henning-krause got the logs. We're investigating.

**Comment by ShreyasRmsft** [comment]
@puagarwa

**Comment by henning-krause** [comment]
@puagarwa, any news on this one?

**Comment by henning-krause** [comment]
@abhishkk, thanks for the hint. I elimated all duplicates, but I still get this error. I sent new logs to the mail address provided above.

Regarding duplicate tests, I have these questions:

- We do have some tests with the same name which are located in different assemblies. Is this a problem?
- We have a test suite with an abstract base class containing a set of tests. Two test classed are derived from this class, alter the behavior of the tests in the base class and add new tests on top of this. Would this be a problem?

**Comment by abhishkk** [comment]
@henning-krause 

> We do have some tests with the same name which are located in different assemblies. Is this a problem?

> We have a test suite with an abstract base class containing a set of tests. Two test classed are derived from this class, alter the behavior of the tests in the base class and add new tests on top of this. Would this be a problem?

Until Namespace.ClassName.MethodName is different, it should create any issues.

We got your new logs. We are looking into it.

**Comment by henning-krause** [comment]
@abhishkk, any news on this?

**Comment by abhishkk** [comment]
@henning-krause Apologies, issue got delayed. We checked that now there are no duplicates anymore. Even though many test methods have same names but their fully qualified name is different. 

We are trying to root cause this issue that why publish is failing. Can you try the run without re-run and share the debug logs?

Also are there any specific tests which you added, after which you are seeing this issue. This will help us in finding the root cause quickly.

**Comment by henning-krause** [comment]
@abhishkk, I think the last Build I sent you didn't have re-runs enabled, just impact analysis. So to be clear, you want a log with impact analysis enabled, but without re-running failed tests?

On a side node, we do have test cases using the [DataRow] attribute which do some case-sensitivity tests. So the method signature would be something this:

```cs
                [TestMethod]
		[DataRow("example.com", "EXAMPLE.COM")]
		[DataRow("EXAMPLE.COM", "example.com")]
		[DataRow("example.com", "example.com")]
		[DataRow("EXAMPLE.COM", "EXAMPLE.COM")]
		[DataRow("Example.com", "example.Com")]
		[DataRow("example.Com", "Example.com")]
		public void Test_CaseSensitive(string a, string b)
```

Since the test id seems to contain the parameters from the datarow, and SQL server by default is not case-sensitive, might this be an issue?

**Comment by nigurr** [comment]
@henning-krause currently we don't support data driven tests with re-run. This feature will be available with TFS 2019 and VS 15.8 only.

**Comment by nigurr** [comment]
@henning-krause Sorry missed that. I looked into the issue. We have an active bug which will address this issue.
Given the bug fix impact, the fix will roll out in 3-4 weeks. 

Thanks for your support in sharing this and helping us to improve the product.
